### ISSUE TITLE
EC2 Instance Profile Credentials Not Registering

### ISSUE THREAD

**Issue by seanherron** [humanmade/S3-Uploads#103]
I have an EC2 instance I've provisioned with an instance profile with the following statements:

```
"Statement": [
                {
                  "Sid": "Stmt1461604304103",
                  "Action": [
                    "s3:ListBucket"
                  ],
                  "Effect": "Allow",
                  "Resource": {"Fn::Join": ["", ["arn:aws:s3:::", { "Ref" : "S3Bucket" }]]}
                },
                {
                  "Sid": "Stmt1461604328103",
                  "Action": [
                    "s3:GetObject",
                    "s3:PutObject",
                    "s3:ListObject"
                  ],
                  "Effect": "Allow",
                  "Resource": {"Fn::Join": ["", ["arn:aws:s3:::", { "Ref" : "S3Bucket" }, "/*"]]}
                }
              ]
```

I can use the aws cli to interact with the bucket successfuly from the instance. When running `wp s3-uploads verify`, however, I get the following error:

```
<Error><Code>InvalidAccessKeyId</Code><Message>The AWS Access Key Id you provided does not exist in our records.</Message><AWSAccessKeyId>S3_UPLOADS_KEY</AWSAccessKeyId>
```

It appears that S3-Uploads is attempting to just pass `S3_UPLOADS_KEY` as the key rather than grabbing the appropriate instance metadata. Here's the settings in `wp-config.php`:

```
define('S3_UPLOADS_BUCKET', '<< bucket name>>' );
define('S3_UPLOADS_USE_INSTANCE_PROFILE', True);
define('S3_UPLOADS_REGION', '<<my region>>');
```

Any ideas?

**Comment by rmccue** [comment]
The way this works internally is handled entirely via the AWS SDK. If you don't set the key/secret constants, S3 Uploads [doesn't pass the `credentials` config into the SDK](https://github.com/humanmade/S3-Uploads/blob/27ab0a4598d809f74f5ac2fa74d8546ef207e567/inc/class-s3-uploads.php#L163-L166). (The `S3_UPLOADS_USE_INSTANCE_PROFILE` constant is just to confirm with S3 Uploads that you're doing this intentionally.)

The error there seems to indicate that the key is getting passed in somehow though, which is strange. Maybe double-check that `S3_UPLOADS_KEY` isn't defined anywhere?

(I'm not entirely sure on this, however @joehoyle is away, so a more authoritative response may be delayed.)

**Comment by seanherron** [comment]
Thanks @rmccue. I can't find anywhere that `S3_UPLOADS_KEY` would be defined - this is on a fresh machine with no other AWS integration running on it. FWIW, I also tried defining `AWS_ACCESS_KEY_ID` as an environmental variable, and that didn't change the fact that `S3_UPLOADS_KEY` seems to be the only value that is passed on to the SDK.

**Comment by rmccue** [comment]
Looks like this should be fixed by #104; the undefined constant was causing the key to be passed as literally `'S3_UPLOADS_KEY'`. Try updating and see if that helps?

**Comment by joehoyle** [comment]
Closing as inactive.
